patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModBrowser.cs

```diff
@@ -501,6 +504,64 @@ internal static async Task<bool> DownloadMods(IEnumerable<ModDownloadItem> mods,
 		}
 	}
 
+	internal static void DownloadModsNoUI(IEnumerable<ModDownloadItem> mods)
+	{
+		var set = mods.ToHashSet();
+		Interface.modBrowser.SocialBackend.GetDependenciesRecursive(set); // Causes a bit of lag
+
+		var fullList = ModDownloadItem.NeedsInstallOrUpdate(set).ToList();
+		if (!fullList.Any())
+			return;
+
+		lock (DownloadQueue) {
+			foreach (var item in fullList) {
+				DownloadQueue.Enqueue(item);
+
+				var notification = new ModDownloadNotification(item);
+				InGameNotificationsTracker.AddMenuNotification(notification);
+				item.notification = notification;
+			}
+		}
+
+		lock (Interface.modBrowser.modSlugsToUpdateInstallInfo) {
+			Interface.modBrowser.modSlugsToUpdateInstallInfo.UnionWith(fullList.Select(x=>x.ModName));
+		}
+
+		if (DownloadTask == null || DownloadTask.IsCompleted) {
+			DownloadTask = Task.Run(DownloadTaskTask);
```

Comment:
You have a race condition here. When `!DownloadTask.IsCompleted` you assume that you don't need to start it again, because it will keep processing `DownloadQueue`, but it could have already exited the `while (DownloadQueue.Count > 0)` loop but not yet reached `IsCompleted` state.

What you want is a producer-consumer channel with only one consumer, that's automatically 'woken' when an item is added to the queue. There are a few patterns on how to do this without race conditions, but by far the easiest is to use https://learn.microsoft.com/en-us/dotnet/core/extensions/channels

An ad-hoc option is to form a task chain, where you start with `DownloadTask = Task.Completed` and when you add an item to the queue, you do `DownloadTask = DownloadTask.ContinueWith(_ => Download(mod))`